### PR TITLE
[RyuJIT/ARM32] Support for CFI unwind info

### PR DESCRIPTION
**Pull request by sergign60**

### CODE REVIEW

**Comment by sergign60** [comment]
@jkotas @BruceForstall @@dotnet/jit-contrib
CC: @Dmitri-Botcharnikov @BredPet @alpencolt
please review

**Comment by sergign60** [comment]
@dotnet-bot test Ubuntu armlb Cross Release Build please

**Comment by sergign60** [comment]
@jkotas @BruceForstall @dotnet/jit-contrib
CC: @Dmitri-Botcharnikov @BredPet @alpencolt
please review

**Comment by sergign60** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please

**Comment by sergign60** [comment]
https://github.com/dotnet/llilc/pull/1011

**Comment by sergign60** [comment]
https://github.com/dotnet/coreclr/pull/2868

**Comment by sergign60** [comment]
https://github.com/dotnet/corert/pull/752

**Comment by sergign60** [comment]
https://github.com/dotnet/corert/pull/4755

**Comment by sergign60** [comment]
https://github.com/dotnet/corert/issues/4626

File: src/jit/unwindarm64.cpp [hunk 1]
```diff
@@ -18,6 +18,15 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
 #if defined(_TARGET_ARM64_)
 
+#if defined(_TARGET_UNIX_)
+int Compiler::mapRegNumToDwarfReg(regNumber reg)
+{
+    int dwarfReg = DWARF_REG_ILLEGAL;
+
+    return dwarfReg;
```
Comment (BruceForstall) [inline src/jit/unwindarm64.cpp]: Is this unimplemented? Should it be:
```
NYI("CFI codes");
```
?

File: src/jit/unwindx86.cpp [hunk 2]
```diff
@@ -20,6 +20,15 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 #error "This should be included only for x86"
 #endif // _TARGET_X86_
 
+#if defined(UNIX_AMD64_ABI) || defined(_TARGET_UNIX_)
+int Compiler::mapRegNumToDwarfReg(regNumber reg)
+{
+    int dwarfReg = DWARF_REG_ILLEGAL;
+
+    return dwarfReg;
```
Comment (BruceForstall) [inline src/jit/unwindx86.cpp]: Is this unimplemented? Should it be:
```
NYI("CFI codes");
```
?

File: src/jit/unwindarm.cpp [hunk 3]
```diff
@@ -16,6 +16,169 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 #pragma hdrstop
 #endif
 
+#if defined(_TARGET_UNIX_)
+
+#if defined(_TARGET_ARM_)
```
Comment (BruceForstall) [inline src/jit/unwindarm.cpp]: How about reducing the two `#if` to one line?

File: src/jit/unwindarm.cpp [hunk 4]
```diff
@@ -201,11 +401,40 @@ void Compiler::unwindPushMaskFloat(regMaskTP maskFloat)
...
+            if (regBit > maskFloat)
+            {
+                break;
+            }
+
+            if (regBit & maskFloat)
+            {
+                unwindPushCFI(regNum);
+            }
+        }
```
Comment (BruceForstall) [inline src/jit/unwindarm.cpp]: This loop looks the same as the one for unwindPushMaskInt(). How about creating a function (unwindPushCFIMask()?)

Note that you could optimize the float case by starting at the first float register; currently it's iterating through all the integer registers just to get to the first float reg.

Comment (BruceForstall) [reply src/jit/unwindarm.cpp]: (You could also use the BitScanForward/BitScanForward64 intrinsics, if you're really worried about perf)

File: src/jit/unwindarm.cpp [hunk 5]
```diff
@@ -415,6 +712,98 @@ void Compiler::unwindEmitFunc(FuncInfoDsc* func, void* pHotCode, void* pColdCode
     static_assert_no_msg(FUNC_HANDLER == (FuncKind)CORJIT_FUNC_HANDLER);
     static_assert_no_msg(FUNC_FILTER == (FuncKind)CORJIT_FUNC_FILTER);
 
+#if defined(_TARGET_UNIX_)
+    if (generateCFIUnwindCodes())
+    {
```
Comment (BruceForstall) [inline src/jit/unwindarm.cpp]: Can you extract this out to a `unwindEmitFuncCFI`?

Comment (sergign60) [reply src/jit/unwindarm.cpp]: @BruceForstall Done

File: src/jit/unwindarm.cpp (hunk 4 above)
Comment (sergign60) [reply src/jit/unwindarm.cpp]: @BruceForstall Donw

**Comment by sergign60** [comment]
@BruceForstall Thanks for your notes. I think that I've satisfied all of them

**Comment by sergign60** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please

**Comment by sergign60** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test please

**Comment by sergign60** [comment]
@dotnet-bot test Tizen armel Cross Debug Build please

**Comment by sergign60** [comment]
@dotnet-bot test CentOS7.1 x64 Release Priority 1 Build and Test please

**Comment by sergign60** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build and Test please

**Comment by BruceForstall** [comment]
Ignoring CentOS; it's failing all jobs.

File: src/jit/unwind.cpp [hunk 6]
```diff
@@ -118,6 +118,299 @@ void Compiler::unwindGetFuncLocations(FuncInfoDsc*             func,
...
+        DumpCfiInfo(true /*isHotCode*/, startOffset, endOffset, unwindCodeBytes, (const CFI_CODE* const)pUnwindBlock);
+    }
+#endif // DEBUG
+
+    assert(endOffset <= info.compTotalHotCodeSize);
+
+    eeAllocUnwindInfo((BYTE*)pHotCode, nullptr /* pColdCode */, startOffset, endOffset, unwindCodeBytes, pUnwindBlock,
+                      (CorJitFuncKind)func->funKind);
+
+    if (pColdCode != NULL)
```
Comment (briansull) [inline src/jit/unwind.cpp]: This NULL needs to be changed to nullptr for clang-tidy to pass

Comment (benaadams) [reply src/jit/unwind.cpp]: Added PR https://github.com/dotnet/coreclr/pull/14602

Comment (sergign60) [reply src/jit/unwind.cpp]: @benaadams Thanks!

**Pull request merged**